prov/efa/src/efa_av.c

```diff
@@ -553,90 +540,115 @@ static int efa_av_remove_ah(struct fid_av *av_fid, fi_addr_t *fi_addr,
+	struct efa_av *av;
+
+	av = container_of(av_fid, struct efa_av, util_av.av_fid);
+	for (i = 0; i < count; i++) {
+		ret = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i],
+				       1, flags);
+		if (ret)
+			return ret;
+	}
+
+	if (i < count) {
+		/* something went wrong, so ret cannot be zero */
+		assert(ret);
+		if (av->util_av.eq) {
+			for (; i < count; ++i)
+				ofi_av_write_event(&av->util_av, i, FI_ECANCELED, NULL);
+		}
+	}
+
+	return 0;
+}
+
 static int efa_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
 			 size_t count, uint64_t flags)
 {
 	int ret = 0;
+	int err;
 	size_t i;
 	struct efa_av *av;
 	struct util_av_entry *util_av_entry;
 	struct efa_av_entry *av_entry;
-	struct rxr_peer *peer;
-	struct dlist_entry *ep_list_entry;
+	struct rdm_peer *peer;
+	struct rxr_ep *rxr_ep;
 
 	av = container_of(av_fid, struct efa_av, util_av.av_fid);
-	if (av->ep_type == FI_EP_DGRAM) {
-		for (i = 0; i < count; i++) {
-			ret = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i],
-					       1, flags);
-			if (ret)
-				goto out;
-		}
-		goto out;
-	}
+	if (av->ep_type == FI_EP_DGRAM)
+		return efa_dgram_av_remove(av_fid, fi_addr, count, flags);
+
 	fastlock_acquire(&av->util_av.lock);
+	/* currently multiple EP bind to same av is not supported */
+	rxr_ep = container_of(av->util_av.ep_list.next, struct rxr_ep, util_ep.av_entry);
 	for (i = 0; i < count; i++) {
-		if (fi_addr[i] == FI_ADDR_NOTAVAIL ||
-		    fi_addr[i] > av->count) {
+		if (fi_addr[i] == FI_ADDR_NOTAVAIL || fi_addr[i] > av->count) {
 			ret = -FI_ENOENT;
-			goto release_lock;
+			goto out;
 		}
-		util_av_entry = ofi_bufpool_get_ibuf(
-						av->util_av.av_entry_pool,
-						fi_addr[i]);
+
+		util_av_entry = ofi_bufpool_get_ibuf(av->util_av.av_entry_pool, fi_addr[i]);
 		if (!util_av_entry) {
 			ret = -FI_ENOENT;
-			goto release_lock;
+			goto out;
 		}
+
+		peer = rxr_ep_get_peer(rxr_ep, fi_addr[i]);
+
+		ret = efa_peer_in_use(peer);
+		if (ret)
+			goto out;
+
+		efa_rdm_peer_release(peer);
+
 		/*
-		 * If use_cnt is greater than 1, then just decrement
-		 * the count by 1, without removing the entry.
+		 * Clearing the 3 resources of an av entry:
+		 *
+		 *     address handler (AH), shm_av_entry and util_av_entry.
+		 *
+		 * We will try our best to remove these resources. If releasing one
+		 * resource failed, we will not stop. Instead we save the error code
+		 * and continue to release other resources.
 		 */
-		if (ofi_atomic_get32(&util_av_entry->use_cnt) > 1) {
-			ret = ofi_av_remove_addr(&av->util_av, fi_addr[i]);
-			goto release_lock;
-		}
 		av_entry = (struct efa_av_entry *)util_av_entry->data;
-
-		/* Check if the peer is in use if it is then return */
-		dlist_foreach(&av->util_av.ep_list, ep_list_entry) {
-			peer = efa_ep_get_peer(ep_list_entry, fi_addr[i]);
-			ret = efa_peer_in_use(peer);
-			if (ret)
-				goto release_lock;
+		err = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i], 1,
+				       flags);
+		if (err)
+			ret = err;
+
```

Comment:
Thanks.  I think that in trying to debug something it might be more helpful if you get the first error code (or, as you did, print out a log that gives a better picture as to where the first error occurred).